nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py

```diff
@@ -0,0 +1,82 @@
+import re
+
+from nmdc_schema.migrators.migrator_base import MigratorBase
+
+
+class Migrator(MigratorBase):
+    r"""Migrates a database between two schemas.
+    """
+
+    _from_version = "10.3.0"
+    _to_version = "10.4.0"
+
+    pattern = re.compile(r"\.[0-9]+$")
+
+    def upgrade(self) -> None:
+        r"""Migrates the database from conforming to the original schema, to conforming to the new schema."""
+
+        self.adapter.process_each_document(
+            "nom_analysis_activity_set", [self.populate_alternative_identifiers,self.update_nom_analysis_activity_id]
+        )
+    
+    def is_valid_id(self, s: str) -> bool:
+        r"""
+        Helper function that checks whether a string is a valid ID (has a version number).
+        Reference: https://docs.python.org/3/library/re.html#re.Pattern.search
+
+        >>> m = Migrator()
+        >>> m.is_valid_id("invalid_id")
```

Comment:
I think this will work but next time we should use the materialized patterns generated by running `make all` which generates `project/nmdc_materialized_patterns.yaml` (^(nmdc):wfnom-([0-9][a-z]{0,6}[0-9])-([A-Za-z0-9]{1,})(\.[0-9]{1,})$)

Reply:
Ah, I didn't know about the patterns, thanks for pointing that out!
I can add that pattern to the regex instead.

Reply:
Hmm, when I plug that in to the regex and test it says IDs without version numbers are still valid.

Reply:
Ah, the issue is a period character is a wildcard so it's just seeing that. If I escape it with a backslash it does work as intended.